subsys/mgmt/mcumgr/lib/cmd/os_mgmt/src/os_mgmt.c

```diff
@@ -12,57 +12,72 @@
 #include <debug/object_tracing.h>
 #include <kernel_structs.h>
 #include <util/mcumgr_util.h>
+#include <zcbor_common.h>
+#include <zcbor_encode.h>
+#include <zcbor_decode.h>
+#include <mgmt/mcumgr/buf.h>
 
-#include "tinycbor/cbor.h"
-#include "cborattr/cborattr.h"
 #include "mgmt/mgmt.h"
 #include "os_mgmt/os_mgmt.h"
 #include "os_mgmt/os_mgmt_impl.h"
 
+/* This is passed to zcbor_map_start/end_endcode as a number of
+ * expected "columns" (tid, priority, and so on)
+ * The value here does not affect memory allocation is is used
+ * to predict how big the map may be. If you increase number
+ * of "columns" the taskstat sends you may need to increase the
+ * value otherwise zcbor_map_end_encode may return with error.
+ */
+#define TASKSTAT_COLUMNS_MAX	20
+
 /**
  * Command handler: os echo
  */
 #if CONFIG_OS_MGMT_ECHO
 static int
 os_mgmt_echo(struct mgmt_ctxt *ctxt)
 {
-	char echo_buf[CONFIG_OS_MGMT_ECHO_LENGTH + 1];
-	CborError err;
-
-	const struct cbor_attr_t attrs[2] = {
-		[0] = {
-			.attribute = "d",
-			.type = CborAttrTextStringType,
-			.addr.string = echo_buf,
-			.nodefault = 1,
-			.len = CONFIG_OS_MGMT_ECHO_LENGTH,
-		},
-		[1] = {
-			.attribute = NULL
-		}
-	};
+	struct zcbor_string value;
+	struct zcbor_string key;
+	bool ok;
 
-	echo_buf[0] = '\0';
+	zcbor_state_t *zsd = ctxt->cnbd->zs;
 
-	err = cbor_read_object(&ctxt->it, attrs);
-	if (err != 0) {
-		return mgmt_err_from_cbor(err);
+	if (!zcbor_map_start_decode(zsd)) {
+		return MGMT_ERR_EUNKNOWN;
 	}
 
-	echo_buf[sizeof(echo_buf) - 1] = '\0';
+	do {
+		ok = zcbor_tstr_decode(zsd, &key);
 
-	err = cbor_encode_text_stringz(&ctxt->encoder, "r")				||
-	      cbor_encode_text_stringz(&ctxt->encoder, echo_buf);
+		if (ok) {
+			if (key.len == 1 && *key.value == 'd') {
+				ok = zcbor_tstr_decode(zsd, &value);
+				break;
+			}
 
-	return (err == 0) ? 0 : MGMT_ERR_ENOMEM;
+			ok = zcbor_any_skip(zsd, NULL);
+		}
+	} while (ok);
+
+	if (ok && !zcbor_map_end_decode(zsd)) {
+		return MGMT_ERR_EUNKNOWN;
+	}
+
+	zcbor_tstr_put_lit(&ctxt->cnbe->zs, "r");
+	if (!zcbor_tstr_encode(&ctxt->cnbe->zs, &value)) {
```

Comment:
If a map is provided that doesn't contain the correct echo request data, what happens at this point?

Reply:
The zcbor_str_decode line that decodes echo request into value would fail if it would not be able to match payload to tstr; so we should never reach this line if payload is not right.